library/objective-c/EnvoyHTTPStreamImpl.m

```diff
@@ -252,13 +262,15 @@ - (int)cancel {
   ios_context *context = _nativeCallbacks.context;
   // Step 1: atomically and synchronously prevent the execution of further callbacks other than
   // on_cancel.
-  if (!atomic_exchange(context->canceled, YES)) {
+  if (!atomic_exchange(context->closed, YES)) {
     // Step 2: directly fire the cancel callback.
     ios_on_cancel(context);
     // Step 3: propagate the reset into native code.
     reset_stream(_streamHandle);
     return 0;
   } else {
+    // Propagate reset, but don't worry about callbacks.
```

Comment:
I would expand the "don't worry about callbacks" to mention that another terminal callback has fired to the platform so all we need to do here is propagate the reset to the engine.